File: leetcode/no0200_Number_of_Islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {

    function showMatrix(grid) {

        console.log("--------------->");
        grid.forEach((row) => {
            console.log(row);
        })
        console.log("---------------<");

    }

    function dfs(grid, x, y, n, m) {
        if (x < 0 || y < 0 || x >= n || y >= m || grid[y][x] == '0')
            return;

        grid[y][x] = '0';
        showMatrix(grid);
        dfs(grid, x + 1, y, n, m);
        dfs(grid, x - 1, y, n, m);
        dfs(grid, x, y + 1, n, m);
        dfs(grid, x, y - 1, n, m);
    }

    let m = grid.length;
    if (m == 0) {
        return 0;
    }
    let n = grid[0].length;

    let ans = 0;
    for (let y = 0; y < m; ++y)
        for (let x = 0; x < n; ++x)
            if (grid[y][x] == '1') {
                ++ans;
                dfs(grid, x, y, n, m);
            }

    return ans;

};

let grid = [
    ["1", "1", "1", "1", "0"],
    ["1", "1", "0", "1", "0"],
    ["1", "1", "0", "0", "0"],
    ["0", "0", "0", "0", "0"]
];

numIslands(grid)

File: leetcode/no0200_Number_of_Islands.go
package main

import "fmt"

func numIslands(grid [][]byte) int {

	// "1": 49
	// "0": 48

	var dfs func(grid [][]byte, x int, y int, w int, h int)
	dfs = func(grid [][]byte, x int, y int, w int, h int) {
		if x < 0 || y < 0 || x >= w || y >= h || grid[x][y] == 48 {
			return
		}
		grid[x][y] = 48
		dfs(grid, x-1, y, w, h)
		dfs(grid, x+1, y, w, h)
		dfs(grid, x, y-1, w, h)
		dfs(grid, x, y+1, w, h)
	}

	width := len(grid)
	height := len(grid[0])

	numberOfIslands := 0

	for x := range grid {

		for y := range grid[x] {
			if grid[x][y] == 49 {
				numberOfIslands += 1
			}

			dfs(grid, x, y, width, height)
		}
	}

	return numberOfIslands
}

func main() {

	grid := [][]byte{
		{49, 49, 49, 49, 48},
		{49, 49, 48, 49, 48},
		{49, 49, 48, 48, 48},
		{48, 48, 48, 48, 48},
	}

	result := numIslands(grid)
	fmt.Println("OUTPUT:", result)
}
